File: app/src/main/java/com/sumian/hw/push/PushReceiver.java
package com.sumian.hw.push;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;

import com.blankj.utilcode.util.LogUtils;
import com.sumian.hw.log.LogManager;
import com.sumian.hw.utils.JsonUtil;
import com.sumian.hw.utils.NotificationUtil;
import com.sumian.sd.app.AppManager;

/**
 * Created by jzz
 * on 2017/11/28.
 * <p>
 * desc:
 */

public class PushReceiver extends BroadcastReceiver {

    private static final String ACTION_PUSH = "com.tech.sumian.action.PUSH";
    private static final String EXTRA_DATA_JSON = "com.avos.avoscloud.Data";

    @Override
    public void onReceive(Context context, Intent intent) {
        String action = intent.getAction();
        if (action == null) {
            return;
        }
        switch (action) {
            case ACTION_PUSH:
                onPush(context, intent);
                break;
            default:
                break;
        }
    }

    private void onPush(Context context, Intent intent) {
        Bundle bundle = intent.getExtras();
        if (bundle == null) {
            return;
        }
        String json = bundle.getString(EXTRA_DATA_JSON);
        LogManager.appendPhoneLog("PushReport json: " + json);
        PushData pushData = JsonUtil.fromJson(json, PushData.class);
        if (pushData == null) {
            return;
        }

        // 过滤非当前用户的推送
        String scheme = pushData.getScheme();
        if (!isUserIdValid(scheme)) {
            return;
        }
        String alert = pushData.getAlert();
        if (TextUtils.isEmpty(alert)) {
            LogUtils.d("alert intent is null");
            return;
        }
        Intent notificationIntent = SchemeUtil.Companion.resolveScheme(context, scheme);
        if (notificationIntent == null) {
            LogUtils.d("push intent is null");
            return;
        }
        NotificationUtil.Companion.showNotification(context, alert, notificationIntent);
    }

    private boolean isUserIdValid(String scheme) {
        String userIdStr = SchemeUtil.Companion.getUserIdFromScheme(scheme);
        if (TextUtils.isEmpty(userIdStr)) {
            return false;
        }
        long pushUserId = Long.valueOf(userIdStr);
        long id = AppManager.getAccountViewModel().getUserInfo().getId();
        if (pushUserId != id) {
            LogUtils.d("user id not equal");
            return false;
        }
        return true;
    }
}

File: app/src/main/java/com/sumian/sleepdoctor/notification/push/PushReceiver.kt
package com.sumian.sleepdoctor.notification.push

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.text.TextUtils
import com.blankj.utilcode.util.LogUtils
import com.sumian.sleepdoctor.app.AppManager
import com.sumian.sleepdoctor.utils.NotificationUtil


/**
 * <pre>
 *     @author : Zhan Xuzhao
 *     e-mail : xuzhao.z@sumian.com
 *     time   : 2018/6/8 10:31
 *     desc   :
 *     version: 1.0
 * </pre>
 */
class PushReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context?, intent: Intent?) {
        LogUtils.d(intent?.action)
        if (context == null) return
        if (intent == null) return
        val pushData: PushData = PushDataResolveUtil.getPushData(intent)
                ?: return
        val scheme = pushData.scheme ?: return
        if (!isUserIdValid(scheme)) {
            return
        }
        val notificationIntent = SchemeResolveUtil.schemeResolver(context, scheme)
                ?: return
        val contentText = pushData.alert ?: return
        NotificationUtil.showNotification(context, contentText, notificationIntent)
        LogUtils.d(pushData)
    }

    private fun isUserIdValid(scheme: String): Boolean {
        val userIdStr = SchemeResolveUtil.getUserIdFromScheme(scheme)
        if (TextUtils.isEmpty(userIdStr)) {
            return false
        }
        val pushUserId = Integer.valueOf(userIdStr)
        val id = AppManager.getAccountViewModel().userInfo.id
        if (pushUserId != id) {
            LogUtils.d("user id not equal")
            return false
        }
        return true
    }
}
